script/test

```diff
@@ -1,4 +1,4 @@
 #!/bin/bash
 set -e
-docker build -t whalebrew .
+docker build  --target build -t whalebrew .
 exec docker run --rm whalebrew go test ./...
```

Comment:
A suggestion could be to make a dedicated `test` stage in the Dockerfile; 

```Dockerfile
FROM build AS test
RUN go test ./...
...
```

The final stage would then copy from the _test_ stage, which forces `docker build` to build both the test and build-stages (this is important when using BuildKit, as it skips steps that are not needed for the specified `--target`);

```
FROM alpine
...

COPY --from=test /go/bin/whalebrew /usr/local/bin/whalebrew
```

That way;

- the source is built (stage `build`)
- the source is tested (stage `test`)
- if the `test` fails, the `docker build` will terminate
- if the `test` is successful, the binary (that was built in the `build` stage) is copied to the final image


With the suggestions above included, the Dockerfile would look something like this;

```Dockerfile
ARG DOCKER_VERSION=18.09
ARG GOLANG_VERSION=1.12

FROM docker:${DOCKER_VERSION} AS docker-cli

FROM golang:${GOLANG_VERSION}-alpine AS build-base
ENV GO111MODULE=on 
RUN mkdir -p /go/src/github.com/whalebrew/whalebrew/
WORKDIR /go/src/github.com/whalebrew/whalebrew/

# build-base is required for tests
RUN apk add --no-cache \
    build-base \
    git

CMD ["whalebrew"]
COPY go.mod go.sum ./
RUN go mod download

COPY . .

FROM build-base AS build
RUN go install .

FROM build AS test
RUN go test ./...

FROM alpine AS deploy
ENTRYPOINT ["whalebrew"]
LABEL io.whalebrew.config.volumes '["/var/run/docker.sock:/var/run/docker.sock", "$WHALEBREW_INSTALL_PATH:$WHALEBREW_INSTALL_PATH"]'
LABEL io.whalebrew.config.keep_container_user 'true'
ENV PATH=${PATH}:/whalebrew

COPY --from=docker:18.06 /usr/local/bin/docker /usr/local/bin/docker
COPY --from=test /go/bin/whalebrew /usr/local/bin/whalebrew
```


When building the image and all tests pass;

```
[+] Building 81.1s (19/19) FINISHED                                                                                                                                                                                                                                             
 => [internal] load build definition from Dockerfile                                                                                                                                                                                                                       0.0s
 => => transferring dockerfile: 904B                                                                                                                                                                                                                                       0.0s
 => [internal] load .dockerignore                                                                                                                                                                                                                                          0.0s
 => => transferring context: 2B                                                                                                                                                                                                                                            0.0s
 => [internal] load metadata for docker.io/library/alpine:latest                                                                                                                                                                                                           0.0s
 => [internal] load metadata for docker.io/library/golang:1.12-alpine                                                                                                                                                                                                      0.0s
 => [build-base 1/6] FROM docker.io/library/golang:1.12-alpine                                                                                                                                                                                                             0.0s
 => => resolve docker.io/library/golang:1.12-alpine                                                                                                                                                                                                                        0.0s
 => [internal] helper image for file operations                                                                                                                                                                                                                            1.6s
 => => resolve docker.io/docker/dockerfile-copy:v0.1.9@sha256:e8f159d3f00786604b93c675ee2783f8dc194bb565e61ca5788f6a6e9d304061                                                                                                                                             0.0s
 => => sha256:e8f159d3f00786604b93c675ee2783f8dc194bb565e61ca5788f6a6e9d304061 2.03kB / 2.03kB                                                                                                                                                                             0.0s
 => => sha256:a546a4352bcaa6512f885d24fef3d9819e70551b98535ed1995e4b567ac6d05b 736B / 736B                                                                                                                                                                                 0.0s
 => => sha256:494e63343c3f0d392e7af8d718979262baec9496a23e97ad110d62b9c90d6182 766B / 766B                                                                                                                                                                                 0.0s
 => => sha256:df3b4bed1f63b36992540a09e0d10bd3f9d0b082d50810313841d745d7cce368 898.21kB / 898.21kB                                                                                                                                                                         1.2s
 => => sha256:f7b6696c3fee7264ec4486cebe146a6a98aa8d1e46747843107ff473aada8d56 861.00kB / 861.00kB                                                                                                                                                                         1.0s
 => => extracting sha256:df3b4bed1f63b36992540a09e0d10bd3f9d0b082d50810313841d745d7cce368                                                                                                                                                                                  0.1s
 => => extracting sha256:f7b6696c3fee7264ec4486cebe146a6a98aa8d1e46747843107ff473aada8d56                                                                                                                                                                                  0.1s
 => [internal] load build context                                                                                                                                                                                                                                          0.1s
 => => transferring context: 279.30kB                                                                                                                                                                                                                                      0.0s
 => FROM docker.io/library/docker:18.06                                                                                                                                                                                                                                    0.0s
 => => resolve docker.io/library/docker:18.06                                                                                                                                                                                                                              0.6s
 => => sha256:fd6f98c37d1f9d550097dbd0193927aa0ab4d290bc9f0fe25b52e9f2cec3f2e7 1.73kB / 1.73kB                                                                                                                                                                             0.0s
 => => sha256:a379bde75b742936ac60746178ae4edf5a029e3c85c202e1c68f210bbb9fb459 1.57kB / 1.57kB                                                                                                                                                                             0.0s
 => => sha256:a60402e528d8a8984ccf117c85e6abad81e3a4a6ad8bf16e014c9441c8bfe5cc 3.99kB / 3.99kB                                                                                                                                                                             0.0s
 => [deploy 1/3] FROM docker.io/library/alpine                                                                                                                                                                                                                             0.0s
 => => resolve docker.io/library/alpine:latest                                                                                                                                                                                                                             0.0s
 => [build-base 2/6] RUN mkdir -p /go/src/github.com/whalebrew/whalebrew/                                                                                                                                                                                                  0.7s
 => [build-base 3/6] RUN apk add --no-cache     build-base     git                                                                                                                                                                                                        11.7s
 => [deploy 2/3] COPY --from=docker:18.06 /usr/local/bin/docker /usr/local/bin/docker                                                                                                                                                                                      0.8s
 => [build-base 4/6] COPY go.mod go.sum ./                                                                                                                                                                                                                                 0.7s
 => [build-base 5/6] RUN go mod download                                                                                                                                                                                                                                  51.3s
 => [build-base 6/6] COPY . .                                                                                                                                                                                                                                              0.6s
 => [build 1/1] RUN go install .                                                                                                                                                                                                                                           9.0s
 => [test 1/1] RUN go test ./...                                                                                                                                                                                                                                           5.9s
 => [deploy 3/3] COPY --from=test /go/bin/whalebrew /usr/local/bin/whalebrew                                                                                                                                                                                               0.7s
 => exporting to image                                                                                                                                                                                                                                                     0.2s
 => => exporting layers                                                                                                                                                                                                                                                    0.2s
 => => writing image sha256:b478beb8dc455c8b1893f578d031fac72ff563d49e075cce450783fc708914cc                                                                                                                                                                               0.0s
 => => naming to docker.io/library/whalebrew                                                                                                                                                                                                                               0.0s
```

And on a failure (I modified one test-case to fail);

```[+] Building 14.5s (16/18)                                                                                                                                                                                                                                                      
 => [internal] load .dockerignore                                                                                                                                                                                                                                          0.0s
 => => transferring context: 2B                                                                                                                                                                                                                                            0.0s
 => [internal] load build definition from Dockerfile                                                                                                                                                                                                                       0.0s
 => => transferring dockerfile: 108B                                                                                                                                                                                                                                       0.0s
 => [internal] load metadata for docker.io/library/golang:1.12-alpine                                                                                                                                                                                                      0.0s
 => [internal] load metadata for docker.io/library/alpine:latest                                                                                                                                                                                                           0.0s
 => CACHED [internal] helper image for file operations                                                                                                                                                                                                                     0.0s
 => FROM docker.io/library/docker:18.06                                                                                                                                                                                                                                    1.2s
 => => resolve docker.io/library/docker:18.06                                                                                                                                                                                                                              1.2s
 => [build-base 1/6] FROM docker.io/library/golang:1.12-alpine                                                                                                                                                                                                             0.0s
 => [deploy 1/3] FROM docker.io/library/alpine                                                                                                                                                                                                                             0.0s
 => [internal] load build context                                                                                                                                                                                                                                          0.0s
 => => transferring context: 10.55kB                                                                                                                                                                                                                                       0.0s
 => CACHED [build-base 2/6] RUN mkdir -p /go/src/github.com/whalebrew/whalebrew/                                                                                                                                                                                           0.0s
 => CACHED [build-base 3/6] RUN apk add --no-cache     build-base     git                                                                                                                                                                                                  0.0s
 => CACHED [build-base 4/6] COPY go.mod go.sum ./                                                                                                                                                                                                                          0.0s
 => CACHED [build-base 5/6] RUN go mod download                                                                                                                                                                                                                            0.0s
 => [build-base 6/6] COPY . .                                                                                                                                                                                                                                              0.5s
 => [build 1/1] RUN go install .                                                                                                                                                                                                                                           8.2s
 => ERROR [test 1/1] RUN go test ./...                                                                                                                                                                                                                                     5.7s
------
 > [test 1/1] RUN go test ./...:
#17 4.626 ?   	github.com/whalebrew/whalebrew	[no test files]
#17 4.626 ?   	github.com/whalebrew/whalebrew/client	[no test files]
#17 4.626 ?   	github.com/whalebrew/whalebrew/cmd	[no test files]
#17 5.349 <nil>
#17 5.349 --- FAIL: TestRun (0.00s)
#17 5.349     --- FAIL: TestRun/When_failing_to_get_current_directory (0.00s)
#17 5.349         hooks_test.go:111: 
#17 5.349             	Error Trace:	hooks_test.go:111
#17 5.349             	Error:      	Received unexpected error:
#17 5.349             	            	unable to get current directory: testError
#17 5.349             	Test:       	TestRun/When_failing_to_get_current_directory
#17 5.349 total 20
#17 5.349 drwxr-xr-x    2 root     root          4096 Apr 20 18:44 .
#17 5.349 drwxr-xr-x    1 root     root          4096 Apr 20 18:44 ..
#17 5.349 -rw-r--r--    1 root     root          1750 Apr 20 18:20 hooks.go
#17 5.349 -rw-r--r--    1 root     root          4538 Apr 20 18:44 hooks_test.go
#17 5.349 FAIL
#17 5.349 FAIL	github.com/whalebrew/whalebrew/hooks	0.006s
#17 5.354 ok  	github.com/whalebrew/whalebrew/packages	0.011s
#17 5.354 ?   	github.com/whalebrew/whalebrew/version	[no test files]
------
executor failed running [/bin/sh -c go test ./...]: exit code: 1
```